sdk/src/fee_calculator.rs

```diff
@@ -1,16 +1,105 @@
 use crate::message::Message;
+use log::*;
 
-#[derive(Serialize, Deserialize, Clone, Debug, Default)]
+#[derive(Serialize, Deserialize, PartialEq, Eq, Clone, Debug, Default)]
 #[serde(rename_all = "camelCase")]
 pub struct FeeCalculator {
+    // The current cost of a signature  This amount may increase/decrease over time based on
+    // cluster processing load.
     pub lamports_per_signature: u64,
+
+    // The target cost of a signature when the cluster is operating around target_signatures_per_slot
+    // signatures
+    pub target_lamports_per_signature: u64,
+
+    // Used to estimate the desired processing capacity of the cluster.  As the signatures for
+    // recent slots are fewer/greater than this value, lamports_per_signature will decrease/increase
+    // for the next slot.  A value of 0 disables lamports_per_signature fee adjustments
+    pub target_signatures_per_slot: usize,
+
+    #[serde(skip)]
+    pub min_lamports_per_signature: u64,
+    #[serde(skip)]
+    pub max_lamports_per_signature: u64,
 }
 
 impl FeeCalculator {
-    pub fn new(lamports_per_signature: u64) -> Self {
-        Self {
-            lamports_per_signature,
+    pub fn new(target_lamports_per_signature: u64) -> Self {
+        let base_fee_calculator = Self {
+            target_lamports_per_signature,
+            lamports_per_signature: target_lamports_per_signature,
+            ..FeeCalculator::default()
+        };
+
+        Self::new_derived(&base_fee_calculator, 0)
+    }
+
+    pub fn new_derived(
+        base_fee_calculator: &FeeCalculator,
+        latest_signatures_per_slot: usize,
+    ) -> Self {
+        let mut me = base_fee_calculator.clone();
+
+        if me.target_signatures_per_slot > 0 {
+            // lamports_per_signature can range from 50% to 1000% of
+            // target_lamports_per_signature
+            //
+            // TODO: Are these decent limits?
+            //
+            me.min_lamports_per_signature = std::cmp::max(1, me.target_lamports_per_signature / 2);
+            me.max_lamports_per_signature = me.target_lamports_per_signature * 10;
+
+            // What the cluster should charge at `latest_signatures_per_slot`
+            let desired_lamports_per_signature = std::cmp::min(
```

Comment:
sigh: makes me wish for std::cmp::constrain(min, val, max)...

Reply:
good idea! https://en.wikipedia.org/wiki/Shunt_(electrical)